### DISCUSSION TITLE
Setting custom args for``InitializeParameters``

### DISCUSSION THREAD

**Discussion by sanjeev55** [eclipse-glsp/glsp#963] [Q&A] [answered]
Hello, 

How do set custom arguments during for the ``InitializeParameters``, where exactly do i have to make changes?
Also, is there a way we can read the URL of the request that was initially made to load the diagram in the glsp server side?
Any information would be appreciated. Thank you.

**Comment by tortmayr** [comment, marked as answer]
Hi @sanjeev55,
I'm basing my answer the Java server implementation & Theia integration. If you are using different components (e.g. the node-based Server implementation) please let me know.

> How do set custom arguments during for the InitializeParameters, where exactly do i have to make changes?

On the client (Theia) side you need to create your own `GLSPClientContribution`. If you are extending from the `BaseGLSPClientContribution` you can simply override the `createInitializeOptions()` method and set your custom arguments there.

For an example you can have a look at the [WorkflowGLSPClientContribution](https://github.com/eclipse-glsp/glsp-theia-integration/blob/256b59459dcd4a166204f3f6f480dcb1d1c1e83a/examples/workflow-theia/src/browser/workflow-glsp-client-contribution.ts#L27).

You also need to bind your custom contribution your [frontend module](https://github.com/eclipse-glsp/glsp-theia-integration/blob/256b59459dcd4a166204f3f6f480dcb1d1c1e83a/examples/workflow-theia/src/browser/workflow-frontend-module.ts#L56).

To handle your custom args on the server side you need to extend the `DefaultGLSPServer` and override the `handleInitializeArgs` method. (see [WorkflowGLSPServer](https://github.com/eclipse-glsp/glsp-server/blob/v1.0.0/examples/org.eclipse.glsp.example.workflow/src/org/eclipse/glsp/example/workflow/WorkflowGLSPServer.java)).

and then you need to create and  use  your custom `ServerModule` e.g:
```java
public class WorkflowServerModule extends ServerModule {

   @Override
   protected Class<? extends GLSPServer> bindGLSPServer() {
      return WorkflowGLSPServer.class;
   }
}
```


 >Also, is there a way we can read the URL of the request that was initially made to load the diagram in the glsp server side?
 
The [RequestModelActionHandler](https://github.com/eclipse-glsp/glsp-server/blob/v1.0.0/plugins/org.eclipse.glsp.server/src/org/eclipse/glsp/server/features/core/model/RequestModelActionHandler.java) stores all client options including the file URL in the `GModelState`. You can use the [ClientOptionsUtil](https://github.com/eclipse-glsp/glsp-server/blob/v1.0.0/plugins/org.eclipse.glsp.server/src/org/eclipse/glsp/server/utils/ClientOptionsUtil.java) to retrieve them e.g:
```java
   @Inject
   protected GModelState modelState;

   public void foo() {
      Optional<String> sourceUri = ClientOptionsUtil.getSourceUri(modelState.getClientOptions());
      // do something
   }
```

**Reply by sanjeev55** [reply]
Hi @tortmayr , thank you for your detailed reply. I will try and implement the way you have suggested. 

Regarding the sourceURI, is there a way to retrieve them, if the provided path to the diagram directory does not exist? What i have noticed till now is if the directory does not have a diagram to load, then the client session not initialized hence does not set ``clientId`` as well and the theia application opens with ``No Folder Opened`` in the explorer. In this situation is there a way to retrieve the URL initially used to make the request. 
For example: the initial request URL to load the diagram is: 
``http://localhost:3000/#/path/to/workspace`` 
but the directory ``workspace` does not exist. How do i handle this? I want the server to handle in such a way that if the directory is not there it creates the directory with a dummy file. Is it possible to do so?

Thank you again, really appreciate your reply.

**Reply by tortmayr** [reply]
Ah I see. I think that is something that needs to be handled on Theia side.
Even assuming that it is possible to retrieve the initial URL and pass it to the GLSP server (which definitely cannot be done out of the box) you wouldn`t gain anything from it. The server startup and the Theia`s workspace initialization happen in parallel. So you will end up with no open Workspace in Theia even if the server creates the corresponding files afterwards.

I don`t know if its possible to hook into Theia`s startup process and create non-existing workspace folders if necessary. 
I would recommend to raise your issue in the [Theia community forum](https://community.theia-ide.org/)

**Reply by sanjeev55** [reply]
Thank you for your reponse @tortmayr. I will get in touch with the Theia community.
